packages/typescript-react-native-resolver/src/host.ts

```diff
@@ -168,33 +163,51 @@ export function resolveModuleNames(
   _reusedNames: string[] | undefined,
   _redirectedReference?: ts.ResolvedProjectReference
 ): (ts.ResolvedModuleFull | undefined)[] {
-  const { options, log, allowedExtensions, replaceReactNativePackageName } =
-    context;
-
-  //
-  //  If the containing file is a type file (.d.ts), it can only import
-  //  other type files and JSON files. Also allow .ts files, as some
-  //  modules import as "foo.d" with the intent to resolve to "foo.d.ts".
-  //
-  const allowedExtensionsDts = [ts.Extension.Dts, ts.Extension.Ts];
-  if (options.resolveJsonModule) {
-    allowedExtensionsDts.push(ts.Extension.Json);
-  }
-  const extensions = hasExtension(containingFile, ts.Extension.Dts)
-    ? allowedExtensionsDts
-    : allowedExtensions;
-
+  const { options, log, replaceReactNativePackageName } = context;
   const resolutions: (ts.ResolvedModuleFull | undefined)[] = [];
 
   for (const moduleName of moduleNames) {
     logModuleBegin(log, moduleName, containingFile);
 
-    const module = resolveModuleName(
+    const finalModuleName = replaceReactNativePackageName(moduleName);
+
+    //  First, try to resolve the module to a TypeScript file. Then, fall back
+    //  to looking for a JavaScript file. Finally, if JSON modules are allowed,
+    //  try resolving to one of them.
+    log.log(
+      "Searching for module '%s' with target file type 'TypeScript'",
+      finalModuleName
+    );
+    let module = resolveModuleName(
       context,
-      replaceReactNativePackageName(moduleName),
+      finalModuleName,
       containingFile,
-      extensions
+      ExtensionsTypeScript
```

Comment:
Why do we have to call `resolveModuleName` three times instead of concatenating the extensions in the right order and calling it once? For instance:
```suggestion
      [
        ...ExtensionsTypeScript,
        ...ExtensionsJavaScript,
        ...(options.resolveJsonModule ? ExtensionsJSON : []),
      ]
```

Reply:
I explained in the PR description... making multiple passes matches what TypeScript does internally. I want to be as close to their resolution scheme as I can be, with the addition of RN resolution logic.